File: apps/ios/Exponential/UI/Session/SessionRowTitle.swift
import ExpCore
import ExpUI
import SwiftUI

/// EXP-688: the first line of a coding-session row — state dot, mono issue
/// identifier, issue title.
///
/// ONE view, two call sites: the Agents list row and the steering screen's
/// nav-bar title, which used to say `Live · macbook` and never named the issue
/// it was steering. Extracted so the two cannot drift (Android's
/// `SessionRowTitle.kt` is the twin).
struct SessionRowTitle: View {
    /// Nil for a batch or action run — those have no issue to name.
    let identifier: String?
    let title: String
    let state: CodingSessionDisplayState
    /// EXP-550: the host machine stopped heartbeating — the run is parked, so
    /// the dot goes static neutral instead of pulsing "coding now". The
    /// steering header widens this to every terminal socket state (closed /
    /// ended), which is just as much "not coding now".
    let paused: Bool
    /// Whether the run is CONNECTED and coding right now. The list reads the
    /// synced row alone, so its rows are live by definition; the steering
    /// header knows better — it holds the live phase, and a connecting or
    /// disconnected screen must not pulse a green "coding now" dot over a
    /// caption that says "Connecting…" or "Session ended".
    var live: Bool = true
    /// EXP-848: the agent is inside a TURN right now — the synced
    /// `coding_sessions.agent_busy` in a list, the screen's own working
    /// predicate while steering. Without it every live row pulsed "coding now"
    /// whether the agent was doing anything or not.
    let busy: Bool

    var body: some View {
        HStack(spacing: 6) {
            if !CodingSessionDisplayState.pulses(
                state: state, agentBusy: busy, paused: paused, live: live
            ) {
                Circle()
                    .fill(paused ? DesignTokens.Semantic.neutral : sessionStateColor(state))
                    .frame(width: 9, height: 9)
            } else {
                PulsingLiveDot()
            }
            if let identifier, !identifier.isEmpty {
                Text(identifier)
                    .font(.caption.monospaced())
                    .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                    .lineLimit(1)
            }
            Text(title)
                .font(.subheadline.weight(.medium))
                .foregroundStyle(.white)
                .lineLimit(1)
                .truncationMode(.tail)
        }
    }
}

/// EXP-804: the run's usage wall — `Rate limited · resets in 2h`.
///
/// A quiet amber pill that renders BESIDE the state badge, never instead of
/// it: the wall is ORTHOGONAL to the session state, so a walled run still
/// reads `running`. Without this the two are indistinguishable — the run just
/// goes silent (the 2026-09-09 incident).
///
/// The label itself is `AgentUsagePresentation.blockedBadgeLabel`, locked ×4;
/// nothing here decides wording. Nil label = not blocked = no pill.
struct SessionBlockedBadge: View {
    /// The raw `coding_sessions.blocked` jsonb text off the wire.
    let blocked: String?
    /// Pinned by the tests; production passes the wall clock so the countdown
    /// re-reads whenever the row redraws.
    var now: Date = Date()

    private var label: String? {
        AgentUsagePresentation.blockedBadgeLabel(
            AgentUsagePresentation.parseBlocked(blocked),
            now: now
        )
    }

    var body: some View {
        if let label {
            HStack(spacing: 4) {
                AppIcon(AppIcons.uiClock, size: 10)
                Text(label)
                    .font(.caption2.weight(.semibold))
                    .lineLimit(1)
            }
            .foregroundStyle(DesignTokens.Semantic.yellow)
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(DesignTokens.Semantic.yellow.opacity(0.12), in: Capsule())
            .accessibilityElement(children: .combine)
            .accessibilityLabel(label)
        }
    }
}

/// Static-dot/label tint per parked display state (EXP-194/EXP-214):
/// review green, done blue (the issue-status palette), needs-input amber.
func sessionStateColor(_ state: CodingSessionDisplayState) -> Color {
    switch state {
    case .needsInput: DesignTokens.Semantic.yellow
    case .review: DesignTokens.Semantic.green
    case .done: DesignTokens.Semantic.blue
    case .running: DesignTokens.Semantic.green
    }
}

/// The word beside the dot on a parked row; nil while the run is simply going.
func sessionStateLabel(_ state: CodingSessionDisplayState) -> String? {
    switch state {
    case .needsInput: "Needs input"
    case .review: "Ready for review"
    case .done: "Done"
    case .running: nil
    }
}

/// The title beside the identifier, one rule for the list and the steering
/// header: an issueless run is an action run when it carries its `action_name`
/// snapshot (EXP-253), else a batch run — never "Untitled issue". A
/// single-issue session whose issue row simply hasn't synced yet (or arrived
/// blank) reads "Untitled issue".
func sessionRowTitle(issue: IssueEntity?, session: CodingSessionEntity) -> String {
    if issue == nil, session.issueId == nil {
        return session.actionName ?? "Batch run"
    }
    let title = issue?.title ?? ""
    return title.isEmpty ? "Untitled issue" : title
}

File: apps/android/app/src/main/java/com/exponential/app/ui/session/SessionRowTitle.kt
package com.exponential.app.ui.session

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.exponential.app.data.db.CodingSessionEntity
import com.exponential.app.data.db.IssueEntity
import com.exponential.app.domain.ISSUE_SYNCING_TITLE
import com.exponential.app.domain.batchRunName
import com.exponential.app.domain.pastRunTitle
import com.exponential.app.ui.theme.TextEmphasis

// EXP-688: a coding session's IDENTITY line — status dot, mono identifier,
// what the run is about. The Agents list row and the steering screen's header
// render the SAME composable so the two can't drift: the header used to say
// only "Live · macbook", which never named the issue being worked on.

/**
 * One session's identity line. [dot] is the caller's status dot — the list
 * derives it from the synced row's display state, the steering header from the
 * live phase — everything else is shared.
 */
@Composable
internal fun SessionRowTitle(
    /** The issue's shortcode — null for a non-issue run, which prints none. */
    identifier: String?,
    title: String,
    modifier: Modifier = Modifier,
    dot: @Composable () -> Unit,
) {
    Row(verticalAlignment = Alignment.CenterVertically, modifier = modifier) {
        dot()
        Spacer(Modifier.width(12.dp))
        if (identifier != null) {
            Text(
                identifier,
                style = MaterialTheme.typography.labelMedium,
                fontFamily = FontFamily.Monospace,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                maxLines = 1,
            )
            Spacer(Modifier.width(8.dp))
        }
        Text(
            title,
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.Medium,
            color = MaterialTheme.colorScheme.onSurface,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f, fill = false),
        )
    }
}

/** EXP-874: the issue shortcode for an issue run; EXP-876: a batch's
 *  `EXP-874 +2`, off the issues it covers ([batchIssues]; empty on a surface
 *  that joins none). An action/chat run — and an issue run whose issue hasn't
 *  synced — prints no identifier. */
internal fun sessionRowIdentifier(
    issue: IssueEntity?,
    session: CodingSessionEntity? = null,
    batchIssues: List<IssueEntity> = emptyList(),
): String? {
    if (issue != null) return issue.identifier
    if (session == null || session.issueId != null || session.actionName != null) return null
    return batchRunName(session, batchIssues).identifier
}

/**
 * What the run is about: the issue's title, an action run's `action_name`
 * snapshot (EXP-253), a chat run's own auto-title (EXP-905), else the batch's
 * own name (EXP-876: its first covered issue's title, else "Batch run") — and,
 * for an issue-scoped session whose issue genuinely hasn't landed yet,
 * [ISSUE_SYNCING_TITLE].
 *
 * EXP-968: the live rows and the Recent ones are named by the SAME rule —
 * this used to carry a second copy of it that said "Issue not synced yet"
 * where every other client (and the Agent page one line below) said
 * "Issue syncing…".
 */
internal fun sessionRowTitle(
    session: CodingSessionEntity,
    issue: IssueEntity?,
    batchIssues: List<IssueEntity> = emptyList(),
): String = pastRunTitle(session, issue, batchIssues)
